theano/tensor/nnet/abstract_conv.py

```diff
@@ -1928,35 +1941,43 @@ def perform(self, node, inp, out_):
             topgrad = new_topgrad
 
-        axes_order = (1, 0) + tuple(range(2, self.convdim + 2))
-        flip_filters = ((slice(None), slice(None)) +
-                        (slice(None, None, -1),) * self.convdim)
-        topgrad = topgrad.transpose(axes_order)[flip_filters]
-        img = img.transpose(axes_order)
-
         if self.unshared is True:
+            flip_kern = ((slice(None), slice(None)) +
+                         (slice(None, None, -1),) * 4)
             kern_shape = (topgrad.shape[1], img.shape[1],
                           topgrad.shape[2], topgrad.shape[3],
-                          topgrad.shape[2] - img.shape[2] + 1,
-                          topgrad.shape[3] - img.shape[3] + 1)
+                          img.shape[2] - topgrad.shape[2] + 1,
+                          img.shape[3] - topgrad.shape[3] + 1)
             kern = np.zeros((kern_shape), dtype=topgrad.dtype)
 
-            for row in range(0, topgrad.shape[2]):
-                for col in range(0, topgrad.shape[3]):
-                    kern[:, :, row, col] = topgrad[:, :, row, col] * \
-                        img[:, :, row:row + kern.shape[4],
-                            col:col + kern.shape[5]]
+            for b in xrange(img.shape[0]):
+                for n in xrange(topgrad.shape[1]):
+                    for im0 in xrange(img.shape[1]):
+                        for row in xrange(topgrad.shape[2]):
+                            for col in xrange(topgrad.shape[3]):
+                                kern[n, im0, row, col, ...] += topgrad[b, n, row, col] * \
+                                    img[b, im0, row:row + kern.shape[4],
+                                        col:col + kern.shape[5]]
+
+            if any(self.filter_dilation[i] > 1 for i in range(self.convdim)):
+                kern = kern[(slice(None),) * 4 +
+                            tuple(slice(None, None, self.filter_dilation[i])
+                                  for i in range(self.convdim))]
         else:
+            axes_order = (1, 0) + tuple(range(2, self.convdim + 2))
+            flip_topgrad = flip_kern = ((slice(None), slice(None)) +
+                                        (slice(None, None, -1),) * self.convdim)
+            topgrad = topgrad.transpose(axes_order)[flip_topgrad]
+            img = img.transpose(axes_order)
             kern = self.conv(img, topgrad, mode="valid")
+            kern = kern.transpose(axes_order)
 
-        if any(self.filter_dilation[i] > 1 for i in range(self.convdim)):
-            kern = kern[(slice(None), slice(None)) +
-                        tuple(slice(None, None, self.filter_dilation[i])
-                              for i in range(self.convdim))]
+            if any(self.filter_dilation[i] > 1 for i in range(self.convdim)):
+                kern = kern[(slice(None), slice(None)) +
+                            tuple(slice(None, None, self.filter_dilation[i])
+                                  for i in range(self.convdim))]
```

Comment:
But the kernel has a different number of dimensions, so don't we have to account for that?